src/main/java/uk/ac/cam/cl/dtg/segue/quiz/PgQuestionAttempts.java

```diff
@@ -227,14 +227,61 @@ public void registerQuestionAttempt(final Long userId, final String questionPage
 
                 listOfResponses.add(questionAttempt);
             }
-            
+
             return mapOfQuestionAttemptsByPage;
         } catch (SQLException e) {
             throw new SegueDatabaseException("Postgres exception", e);
         } catch (IOException e) {
             throw new SegueDatabaseException("Exception while parsing json", e);
         }
     }
+
+    @Override
+    public List<String> getMostRecentQuestionPageAttempts(final Long userId, final Integer limit)
+            throws SegueDatabaseException {
+        PreparedStatement pst;
+        try (Connection conn = database.getDatabaseConnection()) {
+            pst = conn.prepareStatement("SELECT SUBSTRING(question_id, 1, STRPOS(question_id, '|') - 1) AS question_page_id, MAX(\"timestamp\") AS latest_timestamp " +
+                    "FROM question_attempts WHERE user_id = ? " +
+                    "GROUP BY question_page_id " +
+                    "ORDER BY latest_timestamp DESC LIMIT ?");
+            pst.setLong(1, userId);
+            pst.setInt(2, limit);
+
+            ResultSet results = pst.executeQuery();
+            List<String> questionPageIds = new ArrayList<>();
+            while (results.next()) {
+                questionPageIds.add(results.getString("question_page_id"));
+            }
+            return questionPageIds;
+        } catch (SQLException e) {
+            throw new SegueDatabaseException("Postgres exception", e);
+        }
+    }
+
+    @Override
+    public List<String> getUnsolvedQuestions(final Long userId)
+            throws SegueDatabaseException {
+        PreparedStatement pst;
+        try (Connection conn = database.getDatabaseConnection()) {
+            // Perform nested query to extract the page ids with at least one question which has been attempted but never solved
+            pst = conn.prepareStatement("SELECT SUBSTRING(question_id, 1, STRPOS(question_id, '|') - 1) AS question_page_id " +
```

Comment:
Unfortunately, this won't actually find all questions that are not solved. It is perfectly possible to have got parts A, B and C correct, and just not have attempted Part D. That question will show as incomplete on the My Progress page, but this will not find it, since there is no attempt in the database for Part D. The database does not know how many question parts a question has, and so you have to use `GameManager::getAllMarkableQuestionPartsDFSOrder` in order to know what parts there are to a question. I can't see an easy way of fixing this without rewriting it entirely, and since we need all the attempts to do the augmenting we don't need this additional database query :/